File: backtracking/combination_sum_2/main.go
package main

import (
  "fmt"
  "sort"
)

func combinationSum2(candidates []int, target int) [][]int {
  var res [][]int

  sort.Ints(candidates)

  var backtrack func(i, curSum int, subset[]int)

  backtrack = func(i, curSum int, subset []int) {
    if curSum == target {
      tmp := make([]int, len(subset))
      copy(tmp, subset)
      res = append(res, tmp)
      return
    } else if i > len(candidates)-1 || curSum > target {
      return
    }

    subset = append(subset, candidates[i])
    backtrack(i+1, curSum+candidates[i], subset)
    subset = subset[:len(subset)-1]

    for i+1 < len(candidates) && candidates[i] == candidates[i+1] {
      i++
    }

    backtrack(i+1, curSum, subset)
  }

  backtrack(0, 0, []int{})

  return res
}

func main() {
  //candidates := []int{10,1,2,7,6,1,5}
  candidates := []int{2,5,2,1,2}
  targetInt := 5
  res := combinationSum2(candidates, targetInt)
  fmt.Println(res)
}

File: backtracking/combination_sum_2/main.py
'''
You are given an array of integers candidates, which may contain duplicates, and a target integer target. Your task is to return a list of all unique combinations of candidates where the chosen numbers sum to target.

Each element from candidates may be chosen at most once within a combination. The solution set must not contain duplicate combinations.

You may return the combinations in any order and the order of the numbers in each combination can be in any order.
'''

class Solution:
    def combinationSum2(self, candidates: list[int], target: int) -> list[list[int]]:
        res = []

        candidates.sort()

        def backtrack(i: int, subset: list[int], cur_sum: int):
            if cur_sum == target:
                res.append(subset.copy())
                return
            elif i >= len(candidates) or cur_sum > target:
                return

            # Pick
            subset.append(candidates[i])
            backtrack(i+1, subset, cur_sum + candidates[i])
            subset.pop()

            # Skip
            j = i + 1
            while j < len(candidates) and candidates[j] == candidates[i]:
                j += 1

            backtrack(j, subset, cur_sum)

        backtrack(0, [], 0)
        return res


candidates = [9,2,2,4,6,1,5]
target = 8

sol = Solution()

print(sol.combinationSum2(candidates, target))
